enhancements/teamsync-with-oidc.md

```diff
@@ -0,0 +1,194 @@
+
+## Open Questions [optional]
+
+1. When team synchronization is disabled from the web UI, do we remove members(synced through OIDC group) from the team or leave it at current state.
+
+## Summary
+
+Quay currently supports team synchronization with LDAP groups to allow flexible permission assignments to user groups defined in LDAP. However, lately there is an increased adaption towards OIDC(OpenID Connect)-based identity providers.
+This feature enables users to leverage an OIDC provider that supports user groups to synchronize them with teams in Quay.
+
+## Motivation
+
+Support external OIDC Authentication that can be plugged into Quay registry along with managing quay teams externally through the identity provider.
+
+### Goals
+
+* Admins can set up an OIDC Authentication system in Quay's config tool
+* Admins can set up team synchronization configuration (like enabling team sync support, sync interval) in Quay's config tool
+* Admin or user(if permitted) can enable team synchronization and configure an OIDC group to sync with, from the Teams page in the UI
+* User's teams are synchronised with OIDC groups only after the user logs into Quay
+* Teams that are synced with OIDC groups, are regularly updated based on the configured `resycn_interval`
+
+### Non-Goals
+
+* User's teams cannot sync with OIDC groups if the user does not log into quay.
+* Team memberships cannot be edited from the UI if team synchronization is enabled.
+
+## Design Details
+
+The workflow of the current proposal revolves around 3 components of Quay:
+1. Config Tool
+2. UI
+3. Login flow
+
+**Config Tool**
+
+The config for connecting with an external OIDC server looks as below:
+
+```
+  OIDC_LOGIN_CONFIG:
+    CLIENT_ID: <client-id>
+    CLIENT_SECRET: <client-secret>
+    OIDC_SERVER: http://<host_address>/realms/<realm_name>/
+    SERVICE_NAME: <name_of_service>
+    LOGIN_SCOPES: ['openid']
+```
+
+When quay starts, the config-tool verifies if quay was able to successfully connect to the OIDC server. If there is any issue reaching to the server, an error is displayed in the quay container logs.
+
+Team Synchronization details can be set up using the below parameters:
+
+```
+    FEATURE_TEAM_SYNCING: boolean (enables team synchronization with the set up OIDC provider)
+    FEATURE_NONSUPERUSER_TEAM_SYNCING_SETUP: boolean (enables non-superusers to setup team sync)
+```
+
+Quay will start only if config-tool verification passes. Once quay starts, the below workflow kicks in.
+
+**UI**
+
+When the user visits a team page at `/organization/<orgname>/teams/<teamname>`, a user can enable or disable team synchronization from the UI. If team synchronization is enabled, additional details of the OIDC group that needs to be synced with the team is entered.
+Once the group details are configured and verified, the team page enters into a read-only state since the team memberships are now controlled through the OIDC server.
```

Comment:
What happens to users who are already added in the team before sync is enabled? IMO they should be removed since OIDC should be the source of truth

Reply:
IIRC that's the way the team sync worker currently works. Whenever a stale team is re-synced, it should be removing Quay members no longer in the matched LDAP group.

Reply:
Great point, I added in this commit - https://github.com/quay/enhancements/pull/28/commits/59b08162ece24e61719c6f2b062d5d2be6011edf 
Any thoughts on what the behaviour should be when team synchronization is disabled?